### DISCUSSION TITLE
Suggestion: Update password hashing docs to include more options and examples

### DISCUSSION THREAD

**Discussion by dosisod** [fastapi/fastapi#9576] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# This is documentation related, see below
```

### Description

*This is a documentation issue/suggestion in regards to the [OAuth2 with Password (and hashing), Bearer with JWT tokens](https://fastapi.tiangolo.com/tutorial/security/oauth2-jwt/) article.*

The [OWASP Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html) recommends not using the `bcrypt` hash function unless you have to, and instead suggests you use the more secure `scrypt` and `argon2id` functions. One of the primary reasons to not use `bcrypt` is that it limits password length to 72 characters (see below for example), which I bargain a lot of people don't know. At the very least we should mention this in the docs.

<details>
<summary>Code showing that `bcrypt` limits password length to 72 chars</summary>

```python
from passlib.context import CryptContext

ctx = CryptContext(schemes=["bcrypt"])

pw = "x" * 72

hash = ctx.hash(pw)

# Invalid password fails as expected
assert not ctx.verify("not the password", hash)

# Correct password works as expected
assert ctx.verify(pw, hash)

# Passwords longer than 72 chars are truncated, so this works (unexpectedly)
assert ctx.verify(pw + "x", hash)
```
</details>

The `passlib` library does support both [`scrypt`](https://passlib.readthedocs.io/en/stable/lib/passlib.hash.scrypt.html?highlight=scrypt) and [`argon2`](https://passlib.readthedocs.io/en/stable/lib/passlib.hash.argon2.html?highlight=scrypt), so the documentation could be easily updated to explain to users what the different hash functions do, giving them enough information to pick the hash function that best suits their needs.

In addition, the standard library includes `scrypt` support in the [`hashlib` module](https://docs.python.org/3/library/hashlib.html#hashlib.scrypt). This would be a good option for people who don't want to install a 3rd party library just for password hashing. Another reason to not use `passlib` is that it appears to be dead (last release was [3 years ago](https://pypi.org/project/passlib/#history)), which for a security related package is not great. Also, it is throwing a deprecation error whenever you try to import it, since the `crypt` module is deprecated in Python 3.11, and will be removed in Python 3.13. An issue was opened for it 7 months ago [here](https://foss.heptapod.net/python-libs/passlib/-/issues/148), though no progress has been made.

In short, we should give examples of different hashing functions (`bcrypt`, `scrypt`, `argon2`), and let the user pick which suits their needs, as well as suggesting they use the builtin `scrypt` to reduce dependencies.

I would be willing to put in the time writing and documenting this, just thought I would see if this would be a worthy change before I started working on it.

(Sorry if that is a lot of context, there wasn't a good template for documentation related questions/issues, so I just went with this one. Perhaps I should make a documentation discussion template?)

### Operating System

Linux

### FastAPI Version

0.95.2

### Python Version

3.11.3

**Comment by JavierSanchezCastro** [comment]
Good to know! I'm using bcrypt in my project and I didn't know was 72 bytes max. Seems interesting to include the other options

**Reply by JavierSanchezCastro** [reply]
I found this: https://passlib.readthedocs.io/en/stable/lib/passlib.hash.bcrypt.html#security-issues

**Reply by ProOreo** [reply]
I would appreciate an update of the documentation, since for beginners like me the warnings are confusing.
